src/main/scala/cc/factorie/app/nlp/lexicon/AhoCorasick.scala

```diff
@@ -26,196 +16,185 @@ import scala.collection.JavaConversions._
-        if (!constructed) {
-            setTransitions()
-        }
-        var i = 0
-        var curNode = root;
-        while (i < input.length) {
-            val tokenString : String = lemmatizer.lemmatize(input.get(i).string)
-            //logger.log(Level.INFO, "Head = " + head + ", idx = " + index + ", label = " + label)
-            val next = curNode.transitionMap.getOrElse(tokenString,null)
-            if (next != null) {
-                curNode = next;
-                i = i + 1;
-            } else if (curNode != root) {
-                curNode = curNode.failNode;
-            } else {
-                curNode = curNode.failNode;
-                i = i + 1;
-            }
-            if (curNode.getEmit) {
-                //annotate tokens
-                var j = i - 1
-                while (j >= (i - curNode.getEmitDepth)) {
-                    featureFunc(input.get(j)) += tag
-                    j = j - 1
-                }
-            } 
+    found
+  }
+  
+  /** Finds all mentions of the trie phrases in a tokenized input text. */
+  def findMentions(input : Seq[String]) : Set[LexiconMention] = {
+    if (!constructed) {
+      setTransitions()
+    }
+    val mentions : Set[LexiconMention] = new java.util.HashSet[LexiconMention]()
+    var i = 0
+    var curNode = root
+    while (i < input.length) {
+      val head : String = input.get(i)
+      //logger.log(Level.INFO, "Head = " + head + ", idx = " + index + ", label = " + label)
+      val next = curNode.lookupToken(head)
+      if (next != None) {
+        curNode = next.get
+        i = i + 1
+      } else if (curNode != root) {
+        curNode = curNode.failNode
+      } else {
+        curNode = curNode.failNode
+        i = i + 1
+      }
+      if (curNode.getEmit) {
+        for (e <- curNode.getOutputSet) {
+          val strBuffer = new StringBuffer()
+          var j = i - e
+          while (j < i-1) {
+            strBuffer.append(input.get(j))
+            strBuffer.append(sep)
+            j = j + 1
+          }
+          strBuffer.append(input.get(j))
+          mentions.add(new LexiconMention(strBuffer.toString,i-e,i))
         }
+      } 
     }
-    
-    /**
-     * Adds a Seq of phrases into the current Trie, and fixes the failure transitions.
-     */
-    def ++=(input : Seq[Seq[String]]) : Unit = {
-        logger.log(Logger.INFO)("Appending to automaton")
-        for (e <- input) {
-            val queue = new Queue[String]()
-            queue ++= e
-            root += queue
+    mentions
+  }
+  
+  /** Tags a Token's features with a specific tag, if it's context forms a phrase in the trie. */
+  def tagMentions(input : Seq[Token], featureFunc : (Token => CategoricalVectorVar[String]), tag : String) : Unit = {
+    if (!constructed) {
+      setTransitions()
+    }
+    var i = 0
+    var curNode = root
+    while (i < input.length) {
```

Comment:
It's not easy to do as the inputs are different and there is mutable state in the loop that needs to be preserved. It might be possible to do it by making a helper method take a function which unwraps the input and a function which tags or collects the mentions, but that would probably be longer and less clear.